compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -4572,6 +4606,262 @@ public void visit(BLangTypedescExpr accessExpr) {
+            BSymbol symbol = symResolver.resolveUnaryOperator(unaryExpr.operator, exprType);
+            if (symbol == symTable.notFoundSymbol) {
+                symbol = symResolver.getUnaryOpsForTypeSets(unaryExpr.operator, exprType);
+            }
+            if (symbol == symTable.notFoundSymbol) {
+                dlog.error(unaryExpr.pos, DiagnosticErrorCode.UNARY_OP_INCOMPATIBLE_TYPES,
+                        unaryExpr.operator, exprType);
+            } else {
+                unaryExpr.opSymbol = (BOperatorSymbol) symbol;
+                actualType = symbol.type.getReturnType();
+            }
+        }
+
+        // Explicitly set actual type
+        if (exprType != symTable.semanticError && unaryExpr.expr.getKind() == NodeKind.NUMERIC_LITERAL &&
+                (referredType.tag == TypeTags.FINITE || referredType.tag == TypeTags.UNION)) {
+            if (referredType.tag == TypeTags.FINITE) {
+                actualType = createFiniteTypeForNumericUnaryExpr(unaryExpr);
+            } else {
+                if (checkForCompatibleFiniteMembersInUnionType(unaryExpr, (BUnionType) referredType)) {
+                    return createFiniteTypeForNumericUnaryExpr(unaryExpr);
+                }
+                // Check for int subtypes
+                LinkedHashSet<BType> intTypesInUnion = getIntSubtypesInUnionType(unaryExpr, (BUnionType) referredType);
+                if (!intTypesInUnion.isEmpty()) {
+                    BType newReferredType = BUnionType.create(null, intTypesInUnion);
+                    BType tempActualType = checkCompatibilityWithConstructedNumericLiteral(unaryExpr, newReferredType);
+                    if (tempActualType != symTable.semanticError) {
+                        return  tempActualType;
+                    }
+                }
+            }
+        } else if (exprType != symTable.semanticError && TypeTags.isIntegerTypeTag(referredType.tag) &&
+                referredType.tag != TypeTags.INT) {
+            BType tempActualType = checkCompatibilityWithConstructedNumericLiteral(unaryExpr, referredType);
+            if (tempActualType != symTable.semanticError) {
+                return  tempActualType;
+            }
+        }
+        return actualType;
+    }
+
+    public BType checkCompatibilityWithConstructedNumericLiteral(BLangUnaryExpr unaryExpr, BType referredType) {
+        BLangNumericLiteral numericLiteral = createNumericLiteralFormUnaryExpr(unaryExpr);
+        // To check value with sign against expected type
+        return clonedExprTypeChecker(numericLiteral, referredType);
+    }
+
+    public LinkedHashSet getIntSubtypesInUnionType(BLangUnaryExpr unaryExpr, BUnionType expectedType) {
+        LinkedHashSet<BType> intTypesInUnion = new LinkedHashSet<>();
+        for (BType type : expectedType.getMemberTypes()) {
+            BType referredType = types.getReferredType(type);
+            if (TypeTags.isIntegerTypeTag(referredType.tag) && type.tag != TypeTags.INT) {
```

Comment:
Shall we add `type.tag != TypeTags.INT` condition first. Also shall we add a comment why we skip `INT` type for int subtype function